sql: TestAmbiguousCommitDueToLeadershipChange failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/b5a19bacc0c61e00b92d404370662bdfdcd7d2d5

Parameters:
```
COCKROACH_PROPOSER_EVALUATED_KV=true
TAGS=stress
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=58004&tab=buildLog

```
I161124 07:38:12.635183 7294 gossip/gossip.go:244  [n?] initial resolvers: []
W161124 07:38:12.635206 7294 gossip/gossip.go:1120  [n?] no resolvers found; use --join to specify a connected node
W161124 07:38:12.655979 7294 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I161124 07:38:12.656733 7294 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161124 07:38:12.658479 7294 server/config.go:443  1 storage engine initialized
I161124 07:38:12.659103 7294 server/node.go:419  [n?] store [n0,s0] not bootstrapped
I161124 07:38:12.664761 7368 storage/replica_proposal.go:351  [s1,r1/1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 411103h38m21.663831493s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-11-24 07:38:12.664590489 +0000 UTC]
I161124 07:38:12.666280 7294 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161124 07:38:12.666369 7294 server/node.go:348  [n?] **** cluster 49b527be-f1ca-4dd8-acfd-80da5bda744c has been created
I161124 07:38:12.666379 7294 server/node.go:349  [n?] **** add additional nodes by specifying --join=127.0.0.1:42410
I161124 07:38:12.667212 7294 base/node_id.go:62  [n1] NodeID set to 1
I161124 07:38:12.667664 7294 storage/store.go:1201  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161124 07:38:12.669577 7294 server/node.go:432  [n1] initialized store [n1,s1]: {Capacity:536870912 Available:536870912 RangeCount:0 LeaseCount:0}
I161124 07:38:12.669604 7294 server/node.go:317  [n1] node ID 1 initialized
I161124 07:38:12.669653 7294 gossip/gossip.go:286  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:42410" > attrs:<> locality:<> 
I161124 07:38:12.669746 7294 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I161124 07:38:12.669783 7294 server/node.go:562  [n1] connecting to gossip network to verify cluster ID...
I161124 07:38:12.669795 7294 server/node.go:582  [n1] node connected via gossip and verified as part of cluster "49b527be-f1ca-4dd8-acfd-80da5bda744c"
I161124 07:38:12.669834 7415 storage/split_queue.go:103  [n1,split,s1,r1/1:/M{in-ax}] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I161124 07:38:12.670726 7415 storage/replica_command.go:2377  [n1,split,s1,r1/1:/M{in-ax}] initiating a split of this range at key /Table/11 [r2]
I161124 07:38:12.672019 7294 server/node.go:367  [n1] node=1: started with [[]=] engine(s) and attributes []
I161124 07:38:12.672054 7294 server/server.go:630  [n1] starting https server at 127.0.0.1:44276
I161124 07:38:12.672061 7294 server/server.go:631  [n1] starting grpc/postgres server at 127.0.0.1:42410
I161124 07:38:12.672069 7294 server/server.go:632  [n1] advertising CockroachDB node at 127.0.0.1:42410
I161124 07:38:12.724506 7458 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:42410} Attrs: Locality:} ClusterID:49b527be-f1ca-4dd8-acfd-80da5bda744c StartedAt:1479973092669798068}
E161124 07:38:12.739628 7415 storage/queue.go:576  [n1,split,s1,r1/1:/{Min-Table/11}] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
I161124 07:38:12.739751 7415 storage/split_queue.go:103  [n1,split,s1,r2/1:/{Table/11-Max}] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
I161124 07:38:12.739789 7415 storage/replica_command.go:2377  [n1,split,s1,r2/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r3]
E161124 07:38:12.740423 7416 storage/queue.go:587  [n1,replicate,s1,r1/1:/{Min-Table/11}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
E161124 07:38:12.740568 7506 storage/queue.go:587  [n1,replicate,s1,r1/1:/{Min-Table/11}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
E161124 07:38:12.745011 7415 storage/queue.go:576  [n1,split,s1,r2/1:/Table/1{1-2}] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
E161124 07:38:12.750449 7416 storage/queue.go:587  [n1,replicate,s1,r2/1:/Table/1{1-2}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
I161124 07:38:12.750530 7415 storage/split_queue.go:103  [n1,split,s1,r3/1:/{Table/12-Max}] splitting at keys [/Table/13/0 /Table/14/0]
I161124 07:38:12.750576 7415 storage/replica_command.go:2377  [n1,split,s1,r3/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r4]
E161124 07:38:12.759941 7415 storage/queue.go:576  [n1,split,s1,r3/1:/Table/1{2-3}] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
I161124 07:38:12.760040 7415 storage/split_queue.go:103  [n1,split,s1,r4/1:/{Table/13-Max}] splitting at keys [/Table/14/0]
I161124 07:38:12.760075 7415 storage/replica_command.go:2377  [n1,split,s1,r4/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r5]
E161124 07:38:12.760108 7416 storage/queue.go:587  [n1,replicate,s1,r3/1:/Table/1{2-3}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
E161124 07:38:12.806240 7416 storage/queue.go:587  [n1,replicate,s1,r4/1:/Table/1{3-4}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
E161124 07:38:12.806675 7416 storage/queue.go:587  [n1,replicate,s1,r5/1:/{Table/14-Max}] purgatory: 0 of 1 store with an attribute matching []; likely not enough nodes in cluster
I161124 07:38:12.866666 7294 gossip/gossip.go:244  [n?] initial resolvers: [127.0.0.1:42410]
W161124 07:38:12.866707 7294 gossip/gossip.go:1122  [n?] no incoming or outgoing connections
W161124 07:38:12.868903 7294 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I161124 07:38:12.869123 7294 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161124 07:38:12.869762 7294 server/config.go:443  1 storage engine initialized
I161124 07:38:12.870062 7294 server/node.go:419  [n?] store [n0,s0] not bootstrapped
I161124 07:38:12.870083 7294 storage/stores.go:296  [n?] read 0 node addresses from persistent storage
I161124 07:38:12.870102 7294 server/node.go:562  [n?] connecting to gossip network to verify cluster ID...
I161124 07:38:12.893462 7617 gossip/client.go:125  [n?] started gossip client to 127.0.0.1:42410
I161124 07:38:12.893807 7636 gossip/server.go:285  [n1] received gossip from unknown node
I161124 07:38:12.894104 7294 server/node.go:582  [n?] node connected via gossip and verified as part of cluster "49b527be-f1ca-4dd8-acfd-80da5bda744c"
I161124 07:38:12.905672 7575 storage/stores.go:312  [n?] wrote 1 node addresses to persistent storage
I161124 07:38:12.906999 7294 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I161124 07:38:12.909271 7294 server/node.go:310  [n?] new node allocated ID 2
I161124 07:38:12.909295 7294 base/node_id.go:62  [n2] NodeID set to 2
I161124 07:38:12.909331 7294 gossip/gossip.go:286  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:57311" > attrs:<> locality:<> 
I161124 07:38:12.909385 7294 server/node.go:367  [n2] node=2: started with [[]=] engine(s) and attributes []
I161124 07:38:12.909412 7294 server/server.go:630  [n2] starting https server at 127.0.0.1:36788
I161124 07:38:12.909425 7294 server/server.go:631  [n2] starting grpc/postgres server at 127.0.0.1:57311
I161124 07:38:12.909436 7294 server/server.go:632  [n2] advertising CockroachDB node at 127.0.0.1:57311
I161124 07:38:12.913092 7194 storage/stores.go:312  [n1] wrote 1 node addresses to persistent storage
I161124 07:38:12.918749 7294 gossip/gossip.go:244  [n?] initial resolvers: [127.0.0.1:42410]
W161124 07:38:12.918774 7294 gossip/gossip.go:1122  [n?] no incoming or outgoing connections
W161124 07:38:12.921054 7294 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I161124 07:38:12.921256 7294 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161124 07:38:12.921984 7294 server/config.go:443  1 storage engine initialized
I161124 07:38:12.922725 7566 server/node.go:543  [n2] bootstrapped store [n2,s2]
I161124 07:38:12.928759 7506 storage/replica_raftstorage.go:459  [n1,replicate,s1,r2/1:/Table/1{1-2}] generated preemptive snapshot e3fc0941 at index 16
I161124 07:38:12.938672 7294 server/node.go:419  [n?] store [n0,s0] not bootstrapped
I161124 07:38:12.938696 7294 storage/stores.go:296  [n?] read 0 node addresses from persistent storage
I161124 07:38:12.938716 7294 server/node.go:562  [n?] connecting to gossip network to verify cluster ID...
I161124 07:38:12.948680 7683 gossip/client.go:125  [n?] started gossip client to 127.0.0.1:42410
I161124 07:38:12.948934 7648 gossip/server.go:285  [n1] received gossip from unknown node
I161124 07:38:12.949449 7294 server/node.go:582  [n?] node connected via gossip and verified as part of cluster "49b527be-f1ca-4dd8-acfd-80da5bda744c"
I161124 07:38:12.949532 7741 storage/stores.go:312  [n?] wrote 1 node addresses to persistent storage
I161124 07:38:12.949568 7741 storage/stores.go:312  [n?] wrote 2 node addresses to persistent storage
I161124 07:38:12.952076 7569 sql/event_log.go:95  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:57311} Attrs: Locality:} ClusterID:49b527be-f1ca-4dd8-acfd-80da5bda744c StartedAt:1479973092909372297}
I161124 07:38:12.958618 7294 kv/dist_sender.go:302  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I161124 07:38:12.959186 7294 server/node.go:310  [n?] new node allocated ID 3
I161124 07:38:12.959206 7294 base/node_id.go:62  [n3] NodeID set to 3
I161124 07:38:12.959237 7294 gossip/gossip.go:286  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:59858" > attrs:<> locality:<> 
I161124 07:38:12.959278 7294 server/node.go:367  [n3] node=3: started with [[]=] engine(s) and attributes []
I161124 07:38:12.959299 7294 server/server.go:630  [n3] starting https server at 127.0.0.1:49769
I161124 07:38:12.959306 7294 server/server.go:631  [n3] starting grpc/postgres server at 127.0.0.1:59858
I161124 07:38:12.959314 7294 server/server.go:632  [n3] advertising CockroachDB node at 127.0.0.1:59858
I161124 07:38:12.964379 7781 storage/stores.go:312  [n1] wrote 2 node addresses to persistent storage
I161124 07:38:12.964571 7782 storage/stores.go:312  [n2] wrote 2 node addresses to persistent storage
I161124 07:38:12.997623 7826 server/node.go:543  [n3] bootstrapped store [n3,s3]
I161124 07:38:13.057517 7506 storage/store.go:3157  [n1,replicate,s1,r2/1:/Table/1{1-2}] streamed snapshot: kv pairs: 10, log entries: 6
I161124 07:38:13.057916 7872 storage/replica_raftstorage.go:621  [n2,s2,r2/?:{-}] applying preemptive snapshot at index 16 (id=e3fc0941, encoded size=7381, 1 rocksdb batches, 6 log entries)
I161124 07:38:13.058327 7872 storage/replica_raftstorage.go:629  [n2,s2,r2/?:/Table/1{1-2}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.059128 7506 storage/replica_command.go:3273  [n1,replicate,s1,r2/1:/Table/1{1-2}] change replicas: read existing descriptor range_id:2 start_key:"\223" end_key:"\224" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I161124 07:38:13.074787 7806 storage/replica.go:2190  [n1,s1,r2/1:/Table/1{1-2}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I161124 07:38:13.075125 7829 sql/event_log.go:95  [n3] Event: "node_join", target: 3, info: {Descriptor:{NodeID:3 Address:{NetworkField:tcp AddressField:127.0.0.1:59858} Attrs: Locality:} ClusterID:49b527be-f1ca-4dd8-acfd-80da5bda744c StartedAt:1479973092959265195}
I161124 07:38:13.075323 7506 storage/replica_raftstorage.go:459  [n1,replicate,s1,r3/1:/Table/1{2-3}] generated preemptive snapshot d7aec751 at index 20
I161124 07:38:13.075799 7506 storage/store.go:3157  [n1,replicate,s1,r3/1:/Table/1{2-3}] streamed snapshot: kv pairs: 25, log entries: 10
I161124 07:38:13.076098 7970 storage/replica_raftstorage.go:621  [n2,s2,r3/?:{-}] applying preemptive snapshot at index 20 (id=d7aec751, encoded size=15822, 1 rocksdb batches, 10 log entries)
I161124 07:38:13.076406 7970 storage/replica_raftstorage.go:629  [n2,s2,r3/?:/Table/1{2-3}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.077530 7991 storage/raft_transport.go:437  [n2] raft transport stream to node 1 established
I161124 07:38:13.077783 7506 storage/replica_command.go:3273  [n1,replicate,s1,r3/1:/Table/1{2-3}] change replicas: read existing descriptor range_id:3 start_key:"\224" end_key:"\225" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I161124 07:38:13.080846 7857 storage/replica.go:2190  [n1,s1,r3/1:/Table/1{2-3}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I161124 07:38:13.081561 7506 storage/replica_raftstorage.go:459  [n1,replicate,s1,r4/1:/Table/1{3-4}] generated preemptive snapshot 262a261a at index 19
I161124 07:38:13.095223 7506 storage/store.go:3157  [n1,replicate,s1,r4/1:/Table/1{3-4}] streamed snapshot: kv pairs: 49, log entries: 9
I161124 07:38:13.095511 7967 storage/replica_raftstorage.go:621  [n3,s3,r4/?:{-}] applying preemptive snapshot at index 19 (id=262a261a, encoded size=16033, 1 rocksdb batches, 9 log entries)
I161124 07:38:13.095870 7967 storage/replica_raftstorage.go:629  [n3,s3,r4/?:/Table/1{3-4}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.096343 7506 storage/replica_command.go:3273  [n1,replicate,s1,r4/1:/Table/1{3-4}] change replicas: read existing descriptor range_id:4 start_key:"\225" end_key:"\226" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I161124 07:38:13.102776 8041 storage/replica.go:2190  [n1,s1,r4/1:/Table/1{3-4}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:3 ReplicaID:2}]
I161124 07:38:13.108611 8058 storage/raft_transport.go:437  [n3] raft transport stream to node 1 established
I161124 07:38:13.117827 7506 storage/replica_raftstorage.go:459  [n1,replicate,s1,r5/1:/{Table/14-Max}] generated preemptive snapshot 17f1b525 at index 11
I161124 07:38:13.118712 7506 storage/store.go:3157  [n1,replicate,s1,r5/1:/{Table/14-Max}] streamed snapshot: kv pairs: 9, log entries: 1
I161124 07:38:13.118977 8064 storage/replica_raftstorage.go:621  [n2,s2,r5/?:{-}] applying preemptive snapshot at index 11 (id=17f1b525, encoded size=504, 1 rocksdb batches, 1 log entries)
I161124 07:38:13.119214 8064 storage/replica_raftstorage.go:629  [n2,s2,r5/?:/{Table/14-Max}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.119967 7506 storage/replica_command.go:3273  [n1,replicate,s1,r5/1:/{Table/14-Max}] change replicas: read existing descriptor range_id:5 start_key:"\226" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I161124 07:38:13.135010 8114 storage/replica.go:2190  [n1,s1,r5/1:/{Table/14-Max}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I161124 07:38:13.135550 7506 storage/replica_raftstorage.go:459  [n1,replicate,s1,r1/1:/{Min-Table/11}] generated preemptive snapshot 6abe3615 at index 206
I161124 07:38:13.157040 7506 storage/store.go:3157  [n1,replicate,s1,r1/1:/{Min-Table/11}] streamed snapshot: kv pairs: 784, log entries: 76
I161124 07:38:13.157459 8045 storage/replica_raftstorage.go:621  [n3,s3,r1/?:{-}] applying preemptive snapshot at index 206 (id=6abe3615, encoded size=157871, 1 rocksdb batches, 76 log entries)
I161124 07:38:13.158720 8045 storage/replica_raftstorage.go:629  [n3,s3,r1/?:/{Min-Table/11}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I161124 07:38:13.159212 7506 storage/replica_command.go:3273  [n1,replicate,s1,r1/1:/{Min-Table/11}] change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\223" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I161124 07:38:13.162233 8162 storage/replica.go:2190  [n1,s1,r1/1:/{Min-Table/11}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:3 ReplicaID:2}]
I161124 07:38:13.162711 7506 storage/queue.go:639  [n1,replicate] purgatory is now empty
I161124 07:38:13.162900 7416 storage/replica_raftstorage.go:459  [n1,replicate,s1,r1/1:/{Min-Table/11}] generated preemptive snapshot 217f59f2 at index 237
I161124 07:38:13.168396 7416 storage/store.go:3157  [n1,replicate,s1,r1/1:/{Min-Table/11}] streamed snapshot: kv pairs: 787, log entries: 107
I161124 07:38:13.171172 8049 storage/replica_raftstorage.go:621  [n2,s2,r1/?:{-}] applying preemptive snapshot at index 237 (id=217f59f2, encoded size=175287, 1 rocksdb batches, 107 log entries)
I161124 07:38:13.172531 8049 storage/replica_raftstorage.go:629  [n2,s2,r1/?:/{Min-Table/11}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=1ms]
I161124 07:38:13.174015 7416 storage/replica_command.go:3273  [n1,replicate,s1,r1/1:/{Min-Table/11}] change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\223" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:3 store_id:3 replica_id:2 > next_replica_id:3 
I161124 07:38:13.191445 8197 storage/replica.go:2190  [n1,s1,r1/1:/{Min-Table/11}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:3}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:3 ReplicaID:2} {NodeID:2 StoreID:2 ReplicaID:3}]
I161124 07:38:13.200672 7416 storage/replica_raftstorage.go:459  [n1,replicate,s1,r2/1:/Table/1{1-2}] generated preemptive snapshot 41087658 at index 19
I161124 07:38:13.202862 7416 storage/store.go:3157  [n1,replicate,s1,r2/1:/Table/1{1-2}] streamed snapshot: kv pairs: 11, log entries: 9
I161124 07:38:13.203260 8199 storage/replica_raftstorage.go:621  [n3,s3,r2/?:{-}] applying preemptive snapshot at index 19 (id=41087658, encoded size=10168, 1 rocksdb batches, 9 log entries)
I161124 07:38:13.203658 8199 storage/replica_raftstorage.go:629  [n3,s3,r2/?:/Table/1{1-2}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.205473 7416 storage/replica_command.go:3273  [n1,replicate,s1,r2/1:/Table/1{1-2}] change replicas: read existing descriptor range_id:2 start_key:"\223" end_key:"\224" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3 
I161124 07:38:13.249223 8173 storage/replica.go:2190  [n1,s1,r2/1:/Table/1{1-2}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I161124 07:38:13.250687 7416 storage/replica_raftstorage.go:459  [n1,replicate,s1,r5/1:/{Table/14-Max}] generated preemptive snapshot 2b63f419 at index 14
I161124 07:38:13.251667 7416 storage/store.go:3157  [n1,replicate,s1,r5/1:/{Table/14-Max}] streamed snapshot: kv pairs: 10, log entries: 4
I161124 07:38:13.251924 8189 storage/replica_raftstorage.go:621  [n3,s3,r5/?:{-}] applying preemptive snapshot at index 14 (id=2b63f419, encoded size=3298, 1 rocksdb batches, 4 log entries)
I161124 07:38:13.252181 8189 storage/replica_raftstorage.go:629  [n3,s3,r5/?:/{Table/14-Max}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.252677 7416 storage/replica_command.go:3273  [n1,replicate,s1,r5/1:/{Table/14-Max}] change replicas: read existing descriptor range_id:5 start_key:"\226" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3 
I161124 07:38:13.268803 8214 storage/replica.go:2190  [n1,s1,r5/1:/{Table/14-Max}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I161124 07:38:13.282228 7416 storage/replica_raftstorage.go:459  [n1,replicate,s1,r4/1:/Table/1{3-4}] generated preemptive snapshot ff8c0764 at index 34
I161124 07:38:13.284205 7416 storage/store.go:3157  [n1,replicate,s1,r4/1:/Table/1{3-4}] streamed snapshot: kv pairs: 75, log entries: 24
I161124 07:38:13.284662 7948 storage/replica_raftstorage.go:621  [n2,s2,r4/?:{-}] applying preemptive snapshot at index 34 (id=ff8c0764, encoded size=39502, 1 rocksdb batches, 24 log entries)
I161124 07:38:13.287343 7948 storage/replica_raftstorage.go:629  [n2,s2,r4/?:/Table/1{3-4}] applied preemptive snapshot in 3ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.288749 7416 storage/replica_command.go:3273  [n1,replicate,s1,r4/1:/Table/1{3-4}] change replicas: read existing descriptor range_id:4 start_key:"\225" end_key:"\226" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:3 store_id:3 replica_id:2 > next_replica_id:3 
I161124 07:38:13.298199 8274 storage/replica.go:2190  [n1,s1,r4/1:/Table/1{3-4}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:3}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:3 ReplicaID:2} {NodeID:2 StoreID:2 ReplicaID:3}]
I161124 07:38:13.301190 7416 storage/replica_raftstorage.go:459  [n1,replicate,s1,r3/1:/Table/1{2-3}] generated preemptive snapshot ccf6d69a at index 23
I161124 07:38:13.303164 7416 storage/store.go:3157  [n1,replicate,s1,r3/1:/Table/1{2-3}] streamed snapshot: kv pairs: 26, log entries: 13
I161124 07:38:13.303495 8290 storage/replica_raftstorage.go:621  [n3,s3,r3/?:{-}] applying preemptive snapshot at index 23 (id=ccf6d69a, encoded size=18609, 1 rocksdb batches, 13 log entries)
I161124 07:38:13.303891 8290 storage/replica_raftstorage.go:629  [n3,s3,r3/?:/Table/1{2-3}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:13.304370 7416 storage/replica_command.go:3273  [n1,replicate,s1,r3/1:/Table/1{2-3}] change replicas: read existing descriptor range_id:3 start_key:"\224" end_key:"\225" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3 
I161124 07:38:13.319343 8223 storage/replica.go:2190  [n1,s1,r3/1:/Table/1{2-3}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I161124 07:38:13.555410 8387 sql/event_log.go:95  [n1,client=127.0.0.1:36814,user=root] Event: "create_database", target: 50, info: {DatabaseName:test Statement:CREATE DATABASE test User:root}
I161124 07:38:13.555840 7415 storage/split_queue.go:103  [n1,split,s1,r5/1:/{Table/14-Max}] splitting at keys [/Table/50/0]
I161124 07:38:13.555882 7415 storage/replica_command.go:2377  [n1,split,s1,r5/1:/{Table/14-Max}] initiating a split of this range at key /Table/50 [r6]
I161124 07:38:13.565970 8387 sql/event_log.go:95  [n1,client=127.0.0.1:36814,user=root] Event: "create_table", target: 51, info: {TableName:test.t Statement:CREATE TABLE test.t (k SERIAL PRIMARY KEY, v INT) User:root}
I161124 07:38:13.579366 7415 storage/split_queue.go:103  [n1,split,s1,r6/1:/{Table/50-Max}] splitting at keys [/Table/51/0]
I161124 07:38:13.579407 7415 storage/replica_command.go:2377  [n1,split,s1,r6/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r7]
I161124 07:38:13.588780 7390 storage/replica_raftstorage.go:459  [n1,s1,r6/1:/Table/5{0-1}] generated Raft snapshot 372358de at index 15
I161124 07:38:13.589913 8540 storage/store.go:3157  [n1,s1,r6/1:/Table/5{0-1}] streamed snapshot: kv pairs: 11, log entries: 5
I161124 07:38:13.590122 8484 storage/replica_raftstorage.go:621  [n3,s3,r6/3:/{Table/50-Max}] applying Raft snapshot at index 15 (id=372358de, encoded size=6168, 1 rocksdb batches, 5 log entries)
I161124 07:38:13.590457 8484 storage/replica_raftstorage.go:629  [n3,s3,r6/3:/Table/5{0-1}] applied Raft snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I161124 07:38:35.881588 7356 gossip/gossip.go:1136  [n1] node has connected to cluster via gossip
I161124 07:38:35.881877 7356 storage/stores.go:312  [n1] wrote 2 node addresses to persistent storage
I161124 07:38:49.265681 7615 gossip/gossip.go:1136  [n2] node has connected to cluster via gossip
I161124 07:38:49.265830 7615 storage/stores.go:312  [n2] wrote 2 node addresses to persistent storage
I161124 07:38:50.164089 7708 gossip/gossip.go:1136  [n3] node has connected to cluster via gossip
I161124 07:38:50.164220 7708 storage/stores.go:312  [n3] wrote 2 node addresses to persistent storage
I161124 07:38:58.736931 7294 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
W161124 07:38:58.737094 8053 storage/raft_transport.go:443  [n1] raft transport stream to node 3 failed: rpc error: code = 13 desc = transport is closing
I161124 07:38:58.737116 7294 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
W161124 07:38:58.737194 7708 gossip/gossip.go:1122  [n3] no incoming or outgoing connections
W161124 07:38:58.737227 8058 storage/raft_transport.go:443  [n3] raft transport stream to node 1 failed: rpc error: code = 13 desc = transport is closing
I161124 07:38:58.737359 7596 vendor/google.golang.org/grpc/transport/http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42410->127.0.0.1:36786: use of closed network connection
I161124 07:38:58.737476 7690 vendor/google.golang.org/grpc/transport/http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42410->127.0.0.1:36789: use of closed network connection
W161124 07:38:58.737608 7986 storage/raft_transport.go:443  [n1] raft transport stream to node 2 failed: rpc error: code = 13 desc = transport is closing
W161124 07:38:58.737641 7991 storage/raft_transport.go:443  [n2] raft transport stream to node 1 failed: rpc error: code = 13 desc = transport is closing
I161124 07:38:58.737652 7474 vendor/google.golang.org/grpc/transport/http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42410->127.0.0.1:36780: use of closed network connection
W161124 07:38:58.737680 7615 gossip/gossip.go:1122  [n2] no incoming or outgoing connections
I161124 07:38:58.738691 7294 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161124 07:38:58.740034 7294 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
	ambiguous_commit_test.go:95: range 7 was not found
```

Attempting to reproduce this in GCE I hit:
```
$ make stress STRESSFLAGS='-p 100 -stderr -maxfails 1' PKG=./pkg/sql TESTS=TestAmbiguousCommitDueToLeadershipChange

...

--- FAIL: TestAmbiguousCommitDueToLeadershipChange (9.37s)
	ambiguous_commit_test.go:110: no active lease
FAIL


ERROR: exit status 1

1196 runs completed, 1 failures, over 1m10s
```

--- a/pkg/base/constants.go
+++ b/pkg/base/constants.go
@@ -32,6 +32,10 @@ const (
 	// it may be aborted by conflicting txns.
 	DefaultHeartbeatInterval = 1 * time.Second
 
+	// DefaultSendNextTimeout is the duration to wait before trying
+	// another replica to send a KV batch.
+	DefaultSendNextTimeout = 500 * time.Millisecond
+
 	// SlowRequestThreshold is the amount of time to wait before considering a
 	// request to be "slow".
 	SlowRequestThreshold = 60 * time.Second

--- a/pkg/base/test_server_args.go
+++ b/pkg/base/test_server_args.go
@@ -66,6 +66,7 @@ type TestServerArgs struct {
 	SSLCertsDir              string
 	TimeSeriesQueryWorkerMax int
 	SQLMemoryPoolSize        int64
+	SendNextTimeout          time.Duration
 
 	// If set, this will be appended to the Postgres URL by functions that
 	// automatically open a connection to the server. That's equivalent to running

--- a/pkg/kv/dist_sender.go
+++ b/pkg/kv/dist_sender.go
@@ -44,9 +44,6 @@ import (
 
 // Default constants for timeouts.
 const (
-	// TODO(bdarnell): make SendNextTimeout configurable.
-	// https://github.com/cockroachdb/cockroach/issues/6719
-	defaultSendNextTimeout   = 500 * time.Millisecond
 	defaultClientTimeout     = 10 * time.Second
 	defaultPendingRPCTimeout = 500 * time.Millisecond
 
@@ -239,7 +236,7 @@ func NewDistSender(cfg DistSenderConfig, g *gossip.Gossip) *DistSender {
 	if cfg.SendNextTimeout != 0 {
 		ds.sendNextTimeout = cfg.SendNextTimeout
 	} else {
-		ds.sendNextTimeout = defaultSendNextTimeout
+		ds.sendNextTimeout = base.DefaultSendNextTimeout
 	}
 	if cfg.SenderConcurrency != 0 {
 		ds.asyncSenderSem = make(chan struct{}, cfg.SenderConcurrency)
@@ -1166,10 +1163,10 @@ func (ds *DistSender) sendToReplicas(
 	defer slowTimer.Stop()
 	slowTimer.Reset(base.SlowRequestThreshold)
 	for {
-		if !transport.IsExhausted() {
+		if timeout, ok := transport.SendNextTimeout(opts.SendNextTimeout); ok {
 			// Only start the send-next timer if we haven't exhausted the transport
 			// (i.e. there is another replica to send to).
-			sendNextTimer.Reset(opts.SendNextTimeout)
+			sendNextTimer.Reset(timeout)
 		}
 
 		select {

--- a/pkg/kv/dist_sender_test.go
+++ b/pkg/kv/dist_sender_test.go
@@ -113,6 +113,15 @@ func (l *legacyTransportAdapter) IsExhausted() bool {
 	return l.called
 }
 
+func (l *legacyTransportAdapter) SendNextTimeout(
+	defaultTimeout time.Duration,
+) (time.Duration, bool) {
+	if l.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
+}
+
 func (l *legacyTransportAdapter) SendNext(ctx context.Context, done chan<- BatchCall) {
 	l.called = true
 	br, err := l.fn(ctx, l.opts, l.replicas, l.args, l.rpcContext)
@@ -1902,6 +1911,15 @@ type slowLeaseHolderTransport struct {
 
 func (t *slowLeaseHolderTransport) IsExhausted() bool {
 	return t.sendCount > t.replicaCount
+}
+
+func (t *slowLeaseHolderTransport) SendNextTimeout(
+	defaultTimeout time.Duration,
+) (time.Duration, bool) {
+	if t.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
 }
 
 func (t *slowLeaseHolderTransport) SendNext(_ context.Context, done chan<- BatchCall) {

--- a/pkg/kv/send_test.go
+++ b/pkg/kv/send_test.go
@@ -111,6 +111,13 @@ func (c *channelSaveTransport) IsExhausted() bool {
 	return c.remaining <= 0
 }
 
+func (c *channelSaveTransport) SendNextTimeout(defaultTimeout time.Duration) (time.Duration, bool) {
+	if c.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
+}
+
 func (c *channelSaveTransport) SendNext(_ context.Context, done chan<- BatchCall) {
 	c.remaining--
 	c.ch <- done
@@ -364,6 +371,13 @@ type firstNErrorTransport struct {
 
 func (f *firstNErrorTransport) IsExhausted() bool {
 	return f.numSent >= len(f.replicas)
+}
+
+func (f *firstNErrorTransport) SendNextTimeout(defaultTimeout time.Duration) (time.Duration, bool) {
+	if f.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
 }
 
 func (f *firstNErrorTransport) SendNext(_ context.Context, done chan<- BatchCall) {

--- a/pkg/kv/transport.go
+++ b/pkg/kv/transport.go
@@ -30,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/opentracing/opentracing-go"
 )
@@ -53,8 +54,9 @@ type batchClient struct {
 	client     roachpb.InternalClient
 	args       roachpb.BatchRequest
 	healthy    bool
-	retried    bool
 	pending    bool
+	retryable  bool
+	deadline   time.Time
 }
 
 // BatchCall contains a response and an RPC error (note that the
@@ -86,6 +88,12 @@ type TransportFactory func(
 type Transport interface {
 	// IsExhausted returns true if there are no more replicas to try.
 	IsExhausted() bool
+
+	// SendNextTimeout returns the timeout after which the next untried,
+	// or retryable, replica may be attempted. Returns a duration
+	// indicating when another replica should be tried, and a bool
+	// indicating whether one should be (if false, duration will be 0).
+	SendNextTimeout(time.Duration) (time.Duration, bool)
 
 	// SendNext sends the rpc (captured at creation time) to the next
 	// replica. May panic if the transport is exhausted. Should not
@@ -148,8 +156,60 @@ type grpcTransport struct {
 	clientPendingMu syncutil.Mutex // protects access to all batchClient pending flags
 }
 
+// IsExhausted returns false if there are any untried replicas
+// remaining. If there are none, it attempts to resurrect replicas
+// which were tried but failed with a retryable error and have a
+// deadline set which has elapsed. If any where resurrected, returns
+// false; true otherwise.
 func (gt *grpcTransport) IsExhausted() bool {
-	return gt.clientIndex == len(gt.orderedClients)
+	gt.clientPendingMu.Lock()
+	defer gt.clientPendingMu.Unlock()
+	if gt.clientIndex < len(gt.orderedClients) {
+		return false
+	}
+	return !gt.maybeResurrectRetryables()
+}
+
+// maybeResurrectRetryables moves already-tried replicas which
+// experienced a retryable error (currently this means a
+// NotLeaseHolderError) into a newly-active state so that they can be
+// retried. Returns true if any replicas were moved to active.
+func (gt *grpcTransport) maybeResurrectRetryables() bool {
+	var resurrect []batchClient
+	for i := 0; i < gt.clientIndex; i++ {
+		if c := gt.orderedClients[i]; !c.pending && c.retryable && timeutil.Since(c.deadline) >= 0 {
+			resurrect = append(resurrect, c)
+		}
+	}
+	for _, c := range resurrect {
+		gt.moveToFrontLocked(c.args.Replica)
+	}
+	return len(resurrect) > 0
+}
+
+// SendNextTimeout returns the default SendOpts.SendNextTimeout if
+// there are any untried replicas in the transport. Otherwise, it
+// returns the earliest deadline of any replicas which experienced
+// retryable errors.
+func (gt *grpcTransport) SendNextTimeout(defaultTimeout time.Duration) (time.Duration, bool) {
+	gt.clientPendingMu.Lock()
+	defer gt.clientPendingMu.Unlock()
+	if gt.clientIndex < len(gt.orderedClients) {
+		return defaultTimeout, true
+	}
+	var deadline time.Time
+	for i := 0; i < gt.clientIndex; i++ {
+		if c := gt.orderedClients[i]; !c.pending && c.retryable {
+			if (deadline == time.Time{}) || c.deadline.Before(deadline) {
+				deadline = c.deadline
+			}
+		}
+	}
+	if (deadline == time.Time{}) {
+		return 0, false
+	}
+	// Returning a negative duration is legal.
+	return deadline.Sub(timeutil.Now()), true
 }
 
 // SendNext invokes the specified RPC on the supplied client when the
@@ -158,7 +218,7 @@ func (gt *grpcTransport) IsExhausted() bool {
 func (gt *grpcTransport) SendNext(ctx context.Context, done chan<- BatchCall) {
 	client := gt.orderedClients[gt.clientIndex]
 	gt.clientIndex++
-	gt.setPending(client.args.Replica, true)
+	gt.setState(client.args.Replica, true /* pending */, false /* retryable */)
 
 	// Fork the original context as this async send may outlast the
 	// caller's context.
@@ -203,7 +263,16 @@ func (gt *grpcTransport) SendNext(ctx context.Context, done chan<- BatchCall) {
 			}
 			return reply, err
 		}()
-		gt.setPending(client.args.Replica, false)
+		// NotLeaseHolderErrors can be retried.
+		var retryable bool
+		if reply != nil && reply.Error != nil {
+			// TODO(spencer): pass the lease expiration when setting the state to
+			// set a more efficient deadline for retrying this replica.
+			if _, ok := reply.Error.GetDetail().(*roachpb.NotLeaseHolderError); ok {
+				retryable = true
+			}
+		}
+		gt.setState(client.args.Replica, false /* pending */, retryable)
 		tracing.FinishSpan(sp)
 		done <- BatchCall{Reply: reply, Err: err}
 	}()
@@ -212,16 +281,22 @@ func (gt *grpcTransport) SendNext(ctx context.Context, done chan<- BatchCall) {
 func (gt *grpcTransport) MoveToFront(replica roachpb.ReplicaDescriptor) {
 	gt.clientPendingMu.Lock()
 	defer gt.clientPendingMu.Unlock()
+	gt.moveToFrontLocked(replica)
+}
+
+func (gt *grpcTransport) moveToFrontLocked(replica roachpb.ReplicaDescriptor) {
 	for i := range gt.orderedClients {
 		if gt.orderedClients[i].args.Replica == replica {
-			// If a call to this replica is active or retried, don't move it.
-			if gt.orderedClients[i].pending || gt.orderedClients[i].retried {
+			// If a call to this replica is active, don't move it.
+			if gt.orderedClients[i].pending {
 				return
 			}
+			// Clear the retryable bit and deadline, as this replica is being made available.
+			gt.orderedClients[i].retryable = false
+			gt.orderedClients[i].deadline = time.Time{}
 			// If we've already processed the replica, decrement the current
 			// index before we swap.
 			if i < gt.clientIndex {
-				gt.orderedClients[i].retried = true
 				gt.clientIndex--
 			}
 			// Swap the client representing this replica to the front.
@@ -242,12 +317,16 @@ func (*grpcTransport) Close() {
 // mutate, but the clients reside in a slice which is shuffled via
 // MoveToFront, making it unsafe to mutate the client through a reference to
 // the slice.
-func (gt *grpcTransport) setPending(replica roachpb.ReplicaDescriptor, pending bool) {
+func (gt *grpcTransport) setState(replica roachpb.ReplicaDescriptor, pending, retryable bool) {
 	gt.clientPendingMu.Lock()
 	defer gt.clientPendingMu.Unlock()
 	for i := range gt.orderedClients {
 		if gt.orderedClients[i].args.Replica == replica {
 			gt.orderedClients[i].pending = pending
+			gt.orderedClients[i].retryable = retryable
+			if retryable {
+				gt.orderedClients[i].deadline = timeutil.Now().Add(gt.opts.SendNextTimeout)
+			}
 			return
 		}
 	}
@@ -297,6 +376,13 @@ type senderTransport struct {
 
 func (s *senderTransport) IsExhausted() bool {
 	return s.called
+}
+
+func (s *senderTransport) SendNextTimeout(defaultTimeout time.Duration) (time.Duration, bool) {
+	if s.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
 }
 
 func (s *senderTransport) SendNext(ctx context.Context, done chan<- BatchCall) {

--- a/pkg/kv/transport_test.go
+++ b/pkg/kv/transport_test.go
@@ -69,39 +69,39 @@ func TestTransportMoveToFront(t *testing.T) {
 		t.Fatalf("expected cient index 0; got %d", gt.clientIndex)
 	}
 
-	// Advance the client index again and verify replica 3 cannot
+	// Advance the client index again and verify replica 3 cann
 	// be moved to front for a second retry.
 	gt.clientIndex++
 	gt.MoveToFront(rd3)
 	verifyOrder([]roachpb.ReplicaDescriptor{rd3, rd1, rd2})
-	if gt.clientIndex != 1 {
-		t.Fatalf("expected cient index 1; got %d", gt.clientIndex)
+	if gt.clientIndex != 0 {
+		t.Fatalf("expected cient index 0; got %d", gt.clientIndex)
 	}
 
 	// Mark replica 2 no longer pending. Should be able to move it.
 	clients[2].pending = false
 	gt.MoveToFront(rd2)
-	verifyOrder([]roachpb.ReplicaDescriptor{rd3, rd2, rd1})
+	verifyOrder([]roachpb.ReplicaDescriptor{rd2, rd1, rd3})
 
 	// Advance client index and move rd1 front; should be no change.
 	gt.clientIndex++
 	gt.MoveToFront(rd1)
-	verifyOrder([]roachpb.ReplicaDescriptor{rd3, rd2, rd1})
+	verifyOrder([]roachpb.ReplicaDescriptor{rd2, rd1, rd3})
 
-	// Advance client index to and and move rd1 to front. Should move
+	// Advance client index and and move rd1 to front. Should move
 	// client index back for a retry.
 	gt.clientIndex++
 	gt.MoveToFront(rd1)
-	verifyOrder([]roachpb.ReplicaDescriptor{rd3, rd2, rd1})
-	if gt.clientIndex != 2 {
-		t.Fatalf("expected cient index 2; got %d", gt.clientIndex)
+	verifyOrder([]roachpb.ReplicaDescriptor{rd2, rd1, rd3})
+	if gt.clientIndex != 1 {
+		t.Fatalf("expected cient index 1; got %d", gt.clientIndex)
 	}
 
-	// Advance client index once more; verify no second retry.
+	// Advance client index once more; verify second retry.
 	gt.clientIndex++
-	gt.MoveToFront(rd1)
-	verifyOrder([]roachpb.ReplicaDescriptor{rd3, rd2, rd1})
-	if gt.clientIndex != 3 {
-		t.Fatalf("expected cient index 3; got %d", gt.clientIndex)
+	gt.MoveToFront(rd2)
+	verifyOrder([]roachpb.ReplicaDescriptor{rd1, rd2, rd3})
+	if gt.clientIndex != 1 {
+		t.Fatalf("expected cient index 1; got %d", gt.clientIndex)
 	}
 }

--- a/pkg/server/config.go
+++ b/pkg/server/config.go
@@ -163,6 +163,11 @@ type Config struct {
 	// Environment Variable: COCKROACH_TIME_UNTIL_STORE_DEAD
 	TimeUntilStoreDead time.Duration
 
+	// SendNextTimeout is the time after which an alternate replica will
+	// be used to attempt sending a KV batch.
+	// Environment Variable: COCKROACH_SEND_NEXT_TIMEOUT
+	SendNextTimeout time.Duration
+
 	// TestingKnobs is used for internal test controls only.
 	TestingKnobs base.TestingKnobs
 
@@ -303,6 +308,7 @@ func MakeConfig() Config {
 		ConsistencyCheckInterval: defaultConsistencyCheckInterval,
 		MetricsSampleInterval:    defaultMetricsSampleInterval,
 		TimeUntilStoreDead:       defaultTimeUntilStoreDead,
+		SendNextTimeout:          base.DefaultSendNextTimeout,
 		EventLogEnabled:          defaultEventLogEnabled,
 		Stores: base.StoreSpecList{
 			Specs: []base.StoreSpec{{Path: defaultStorePath}},
@@ -445,6 +451,7 @@ func (cfg *Config) readEnvironmentVariables() {
 	cfg.ScanInterval = envutil.EnvOrDefaultDuration("COCKROACH_SCAN_INTERVAL", cfg.ScanInterval)
 	cfg.ScanMaxIdleTime = envutil.EnvOrDefaultDuration("COCKROACH_SCAN_MAX_IDLE_TIME", cfg.ScanMaxIdleTime)
 	cfg.TimeUntilStoreDead = envutil.EnvOrDefaultDuration("COCKROACH_TIME_UNTIL_STORE_DEAD", cfg.TimeUntilStoreDead)
+	cfg.SendNextTimeout = envutil.EnvOrDefaultDuration("COCKROACH_SEND_NEXT_TIMEOUT", cfg.SendNextTimeout)
 	cfg.ConsistencyCheckInterval = envutil.EnvOrDefaultDuration("COCKROACH_CONSISTENCY_CHECK_INTERVAL", cfg.ConsistencyCheckInterval)
 }
 

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -198,6 +198,7 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 		Clock:           s.clock,
 		RPCContext:      s.rpcContext,
 		RPCRetryOptions: &retryOpts,
+		SendNextTimeout: s.cfg.SendNextTimeout,
 	}
 	s.distSender = kv.NewDistSender(distSenderCfg, s.gossip)
 	s.registry.AddMetricStruct(s.distSender.Metrics())

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -127,6 +127,9 @@ func makeTestConfigFromParams(params base.TestServerArgs) Config {
 	if params.SQLMemoryPoolSize != 0 {
 		cfg.SQLMemoryPoolSize = params.SQLMemoryPoolSize
 	}
+	if params.SendNextTimeout != 0 {
+		cfg.SendNextTimeout = params.SendNextTimeout
+	}
 	cfg.JoinList = []string{params.JoinAddr}
 	if cfg.Insecure {
 		// Whenever we can (i.e. in insecure mode), use IsolatedTestAddr

--- a/pkg/sql/ambiguous_commit_test.go
+++ b/pkg/sql/ambiguous_commit_test.go
@@ -18,8 +18,10 @@ package sql_test
 
 import (
 	"bytes"
+	"fmt"
 	"sync/atomic"
 	"testing"
+	"time"
 
 	"golang.org/x/net/context"
 
@@ -28,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -47,117 +50,133 @@ import (
 func TestAmbiguousCommitDueToLeadershipChange(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	// Create a command filter which prevents EndTransaction from
-	// returning a response.
-	params := base.TestServerArgs{}
-	committed := make(chan struct{})
-	wait := make(chan struct{})
-	var tableStartKey atomic.Value
-	var responseCount int32
+	for _, requireDistSenderRetry := range []bool{true, false} {
+		t.Run(fmt.Sprintf("retry=%t", requireDistSenderRetry), func(t *testing.T) {
+			// Create a command filter which prevents EndTransaction from
+			// returning a response.
+			params := base.TestServerArgs{}
+			committed := make(chan struct{})
+			wait := make(chan struct{})
+			var tableStartKey atomic.Value
+			var responseCount int32
 
-	// Prevent the first conditional put on table 51 from returning to
-	// waiting client in order to simulate a lost update or slow network
-	// link.
-	params.Knobs.Store = &storage.StoreTestingKnobs{
-		TestingResponseFilter: func(ba roachpb.BatchRequest, br *roachpb.BatchResponse) *roachpb.Error {
-			req, ok := ba.GetArg(roachpb.ConditionalPut)
-			tsk := tableStartKey.Load()
-			if tsk == nil {
-				return nil
+			// Prevent the first conditional put on table 51 from returning to
+			// waiting client in order to simulate a lost update or slow network
+			// link.
+			params.Knobs.Store = &storage.StoreTestingKnobs{
+				TestingResponseFilter: func(ba roachpb.BatchRequest, br *roachpb.BatchResponse) *roachpb.Error {
+					req, ok := ba.GetArg(roachpb.ConditionalPut)
+					tsk := tableStartKey.Load()
+					if tsk == nil {
+						return nil
+					}
+					if !ok || !bytes.HasPrefix(req.Header().Key, tsk.([]byte)) {
+						return nil
+					}
+					// If this is the first write to the table, wait to respond to the
+					// client in order to simulate a retry.
+					if atomic.AddInt32(&responseCount, 1) == 1 {
+						close(committed)
+						<-wait
+					}
+					return nil
+				},
 			}
-			if !ok || !bytes.HasPrefix(req.Header().Key, tsk.([]byte)) {
-				return nil
+			params.SendNextTimeout = 50 * time.Millisecond
+			testClusterArgs := base.TestClusterArgs{
+				ReplicationMode: base.ReplicationAuto,
+				ServerArgs:      params,
 			}
-			// If this is the first write to the table, wait to respond to the
-			// client in order to simulate a retry.
-			if atomic.AddInt32(&responseCount, 1) == 1 {
-				close(committed)
-				<-wait
-			}
-			return nil
-		},
-	}
-	testClusterArgs := base.TestClusterArgs{
-		ReplicationMode: base.ReplicationAuto,
-		ServerArgs:      params,
-	}
-	const numReplicas = 3
-	tc := testcluster.StartTestCluster(t, numReplicas, testClusterArgs)
-	defer tc.Stopper().Stop(context.TODO())
+			const numReplicas = 3
+			tc := testcluster.StartTestCluster(t, numReplicas, testClusterArgs)
+			defer tc.Stopper().Stop(context.TODO())
 
-	sqlDB := sqlutils.MakeSQLRunner(t, tc.Conns[0])
+			sqlDB := sqlutils.MakeSQLRunner(t, tc.Conns[0])
 
-	sqlDB.Exec(`CREATE DATABASE test`)
-	sqlDB.Exec(`CREATE TABLE test.t (k SERIAL PRIMARY KEY, v INT)`)
+			sqlDB.Exec(`CREATE DATABASE test`)
+			sqlDB.Exec(`CREATE TABLE test.t (k SERIAL PRIMARY KEY, v INT)`)
 
-	tableID, err := sqlutils.QueryTableID(tc.Conns[0], "test", "t")
-	if err != nil {
-		t.Fatal(err)
-	}
-	tableStartKey.Store(keys.MakeTablePrefix(tableID))
-
-	// Wait for new table to split & replication.
-	if err := tc.WaitForSplitAndReplication(tableStartKey.Load().([]byte)); err != nil {
-		t.Fatal(err)
-	}
-
-	// Lookup the lease.
-	tableRangeDesc, err := tc.LookupRange(keys.MakeRowSentinelKey(tableStartKey.Load().([]byte)))
-	if err != nil {
-		t.Fatal(err)
-	}
-	leaseHolder, err := tc.FindRangeLeaseHolder(
-		tableRangeDesc,
-		&roachpb.ReplicationTarget{
-			NodeID:  tc.Servers[0].GetNode().Descriptor.NodeID,
-			StoreID: tc.Servers[0].GetFirstStoreID(),
-		})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// In a goroutine, send an insert which will commit but not return
-	// from the leader (due to the command filter we installed on node 0).
-	sqlErrCh := make(chan error, 1)
-	go func() {
-		// Use a connection other than through the node which is the current
-		// leaseholder to ensure that we use GRPC instead of the local server.
-		// If we use a local server, the hanging response we simulate takes
-		// up the dist sender thread of execution because local requests are
-		// executed synchronously.
-		sqlConn := tc.Conns[leaseHolder.NodeID%numReplicas]
-		_, err := sqlConn.Exec(`INSERT INTO test.t (v) VALUES (1)`)
-		sqlErrCh <- err
-		close(wait)
-	}()
-	// Wait until the insert has committed.
-	<-committed
-
-	// Find a node other than the current lease holder to transfer the lease to.
-	for i, s := range tc.Servers {
-		if leaseHolder.StoreID != s.GetFirstStoreID() {
-			if err := tc.TransferRangeLease(tableRangeDesc, tc.Target(i)); err != nil {
+			tableID, err := sqlutils.QueryTableID(tc.Conns[0], "test", "t")
+			if err != nil {
 				t.Fatal(err)
 			}
-			break
-		}
-	}
+			tableStartKey.Store(keys.MakeTablePrefix(tableID))
 
-	// Wait for the error from the pending SQL insert.
-	err = <-sqlErrCh
-	if pqErr, ok := err.(*pq.Error); !ok {
-		t.Errorf("expected ambiguous commit error with correct code; got %v", err)
-	} else {
-		if pqErr.Code != pgerror.CodeStatementCompletionUnknownError {
-			t.Errorf("expected code %q, got %q (err: %s)",
-				pgerror.CodeStatementCompletionUnknownError, pqErr.Code, err)
-		}
-	}
+			// Wait for new table to split & replication.
+			if err := tc.WaitForSplitAndReplication(tableStartKey.Load().([]byte)); err != nil {
+				t.Fatal(err)
+			}
 
-	// Verify a single row exists in the table.
-	var rowCount int
-	sqlDB.QueryRow(`SELECT count(*) FROM test.t`).Scan(&rowCount)
-	if rowCount != 1 {
-		t.Errorf("expected 1 row but found %d", rowCount)
+			// Lookup the lease.
+			tableRangeDesc, err := tc.LookupRange(keys.MakeRowSentinelKey(tableStartKey.Load().([]byte)))
+			if err != nil {
+				t.Fatal(err)
+			}
+			// Loop until we can get the lease holder. We query in the loop to
+			// ensure that the lease is acquired.
+			var leaseHolder roachpb.ReplicationTarget
+			testutils.SucceedsSoon(t, func() error {
+				rows := sqlDB.Query(`SELECT * FROM test.t`)
+				rows.Close()
+				leaseHolder, err = tc.FindRangeLeaseHolder(
+					tableRangeDesc,
+					&roachpb.ReplicationTarget{
+						NodeID:  tc.Servers[0].GetNode().Descriptor.NodeID,
+						StoreID: tc.Servers[0].GetFirstStoreID(),
+					})
+				return err
+			})
+
+			// In a goroutine, send an insert which will commit but not return
+			// from the leader (due to the command filter we installed on node 0).
+			sqlErrCh := make(chan error, 1)
+			go func() {
+				// Use a connection other than through the node which is the current
+				// leaseholder to ensure that we use GRPC instead of the local server.
+				// If we use a local server, the hanging response we simulate takes
+				// up the dist sender thread of execution because local requests are
+				// executed synchronously.
+				sqlConn := tc.Conns[leaseHolder.NodeID%numReplicas]
+				_, err := sqlConn.Exec(`INSERT INTO test.t (v) VALUES (1)`)
+				sqlErrCh <- err
+				close(wait)
+			}()
+			// Wait until the insert has committed.
+			<-committed
+
+			// If requested, wait for two further attempts from the distributed sender.
+			if requireDistSenderRetry {
+				// Wait for twice the send next timeout.
+				time.Sleep(2 * params.SendNextTimeout)
+			}
+
+			// Find a node other than the current lease holder to transfer the lease to.
+			for i, s := range tc.Servers {
+				if leaseHolder.StoreID != s.GetFirstStoreID() {
+					if err := tc.TransferRangeLease(tableRangeDesc, tc.Target(i)); err != nil {
+						t.Fatal(err)
+					}
+					break
+				}
+			}
+
+			// Wait for the error from the pending SQL insert.
+			err = <-sqlErrCh
+			if pqErr, ok := err.(*pq.Error); !ok {
+				t.Errorf("expected ambiguous commit error with correct code; got %v", err)
+			} else {
+				if pqErr.Code != pgerror.CodeStatementCompletionUnknownError {
+					t.Errorf("expected code %q, got %q (err: %s)",
+						pgerror.CodeStatementCompletionUnknownError, pqErr.Code, err)
+				}
+			}
+
+			// Verify a single row exists in the table.
+			var rowCount int
+			sqlDB.QueryRow(`SELECT count(*) FROM test.t`).Scan(&rowCount)
+			if rowCount != 1 {
+				t.Errorf("expected 1 row but found %d", rowCount)
+			}
+		})
 	}
 }

--- a/pkg/storage/client_test.go
+++ b/pkg/storage/client_test.go
@@ -436,6 +436,15 @@ func (t *multiTestContextKVTransport) IsExhausted() bool {
 	return t.idx == len(t.replicas)
 }
 
+func (t *multiTestContextKVTransport) SendNextTimeout(
+	defaultTimeout time.Duration,
+) (time.Duration, bool) {
+	if t.IsExhausted() {
+		return 0, false
+	}
+	return defaultTimeout, true
+}
+
 func (t *multiTestContextKVTransport) SendNext(ctx context.Context, done chan<- kv.BatchCall) {
 	rep := t.replicas[t.idx]
 	t.idx++

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -986,6 +986,14 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
 						newNotLeaseHolderError(&transferLease, r.store.StoreID(), r.mu.state.Desc))
 				}
 
+				// If the lease is in stasis, we can't serve requests until we've
+				// renewed the lease, so we return the channel to block on renewal.
+				// Otherwise, we don't need to wait for the extension and simply
+				// ignore the returned channel (which is buffered) and continue.
+				if status.state == leaseStasis {
+					return r.requestLeaseLocked(ctx, status), nil
+				}
+
 				// Extend the lease if this range uses expiration-based
 				// leases, the lease is in need of renewal, and there's not
 				// already an extension pending.
@@ -997,15 +1005,9 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
 							log.Infof(ctx, "extending lease %s at %s", status.lease, timestamp)
 						}
 						// We had an active lease to begin with, but we want to trigger
-						// a lease extension.
-						llChan := r.requestLeaseLocked(ctx, status)
-						// If the lease is in stasis, we can't serve requests until we've
-						// renewed the lease, so we return the channel to block on renewal.
-						// Otherwise, we don't need to wait for the extension and simply
-						// ignore the returned channel (which is buffered) and continue.
-						if status.state == leaseStasis {
-							return llChan, nil
-						}
+						// a lease extension. We explicitly ignore the returned channel
+						// as we won't block on it.
+						_ = r.requestLeaseLocked(ctx, status)
 					}
 				}
 

